### PR TITLE
[JENKINS-64006] Implement Dynamic PVC for volume pod template

### PR DESCRIPTION
**Pull request by Vlatombe**
The existing logic in DynamicPVCWorkspaceVolume has been moved to a
separate interface DynamicPVC, so that it can be shared with the new
DynamicPVCVolume implementation.

* Renamed `KubernetesPipelineTest#dynamicPVC` as it was ambiguous with the new use case
* Added `KubernetesPipelineTest#dynamicPVCVolume` test

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 1]
```diff
@@ -184,7 +184,7 @@ public Pod build() {
         // Build volumes and volume mounts.
         Map<String, Volume> volumes = new HashMap<>();
         Map<String, VolumeMount> volumeMounts = new HashMap<>();
-
+        String podName = agent != null ? agent.getPodName() : null;
```
Comment (aHenryJard) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:187]: Are you sure it's not an issue that podName can be null ? I wonder if it's not better to put a default name like "undetermined" or fail fast and throw an exception.

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVCVolume/help-requestsSize.html [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Claims, like pods, can request specific quantities of a resource. In this case, the request is for storage. The same resource model applies to both volumes and claims.
+
```
Comment (aHenryJard) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVCVolume/help-requestsSize.html:2]: If I understand correctly getRequestsSizeOrDefault, if this field is empty then default is 10Gi, should be explained to end users. And it gives an example also.
```suggestion
If this field is empty, the default value of 10Gi will be taken.
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java [hunk 3]
```diff
@@ -41,8 +44,26 @@
...
     public abstract String getMountPath();
 
+    // Builds a Volume model with the given name.require podName to generate pvc name
+    public Volume buildVolume(String volumeName, String podName){
+        return buildVolume(volumeName);
+    };
+
     // Builds a Volume model with the given name.
-    public abstract Volume buildVolume(String volumeName);
+    @Deprecated
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java:53]: Some javadoc here would be good IMO. it's unclear why this is throwing `UnsupportedOperationException` and why subclases are granted to have an override of this, unless you come to this PR and see that the method was `abstract` before this change.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java [hunk 4]
```diff
@@ -41,8 +44,26 @@
     // Where to mount this volume in the pod.
     public abstract String getMountPath();
 
+    // Builds a Volume model with the given name.require podName to generate pvc name
+    public Volume buildVolume(String volumeName, String podName){
+        return buildVolume(volumeName);
+    };
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java:50]:
```suggestion
    }
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java [hunk 5]
```diff
@@ -41,8 +44,26 @@
     // Where to mount this volume in the pod.
     public abstract String getMountPath();
 
+    // Builds a Volume model with the given name.require podName to generate pvc name
+    public Volume buildVolume(String volumeName, String podName){
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java:48]:
```suggestion
    public Volume buildVolume(String volumeName, String podName) {
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java [hunk 6]
```diff
@@ -41,8 +44,26 @@
...
+        return buildVolume(volumeName);
+    };
+
     // Builds a Volume model with the given name.
-    public abstract Volume buildVolume(String volumeName);
+    @Deprecated
+    public Volume buildVolume(String volumeName){
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/PodVolume.java:54]:
```suggestion
    public Volume buildVolume(String volumeName) {
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVCVolume.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    public static class DescriptorImpl extends Descriptor<PodVolume> {
+        @Override
+        public String getDisplayName() {
+            return "Dynamic Persistent Volume Claim";
+        }
+
+        @SuppressWarnings("unused") // by stapler
+        @RequirePOST
+        @Restricted(DoNotUse.class) // stapler only
+        public ListBoxModel doFillAccessModesItems(){
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVCVolume.java:119]:
```suggestion
        public ListBoxModel doFillAccessModesItems() {
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVC.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                .withClaimName(pvcName)
+                .withReadOnly(false)
+                .endPersistentVolumeClaim()
+                .build();
+    }
+
+    default PersistentVolumeClaim createPVC(KubernetesClient client, ObjectMeta podMetaData){
+        String namespace = podMetaData.getNamespace();
+        String podName = podMetaData.getName();
+        LOGGER.log(Level.FINE, "Adding volume from pod: {0}/{1}", new Object[] { namespace, podName });
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVC.java:40]: Is this `from` or `for`?

**Review by amuniz — APPROVED** [review]

**Review by aHenryJard — APPROVED** [review]
Looks good

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVC.java (hunk 8 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/DynamicPVC.java]: Rather `for` indeed. I moved the initial code without paying too much attention to the exact wording.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/workspace/DynamicPVCWorkspaceVolume.java [hunk 9]
```diff
@@ -82,72 +63,17 @@ public void setStorageClassName(@CheckForNull String storageClassName) {
...
-                withBlockOwnerDeletion(true).
-                withController(true).
-                withName(podMetaData.getName()).
-                withUid(podMetaData.getUid()).build();
-
-         PersistentVolumeClaim pvc = new PersistentVolumeClaimBuilder()
-                .withNewMetadata()
-                .withName("pvc-" + podMetaData.getName())
-                .withOwnerReferences(ownerReference)
-                .withLabels(DEFAULT_POD_LABELS)
```
Comment (StevenPZChan) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/workspace/DynamicPVCWorkspaceVolume.java:111]: Why removed the labels here? I used the labels to distinguish PVC created by Jenkins, and found it not worked after I upgraded the plugin to the latest version. :cry:

**Pull request merged**